navigation/pose_action_server/src/pose_action_server_ros.cpp

```diff
@@ -0,0 +1,148 @@
+            this, action_name,
+            std::bind(&PoseActionServerNode::handleGoal, this, _1, _2),
+            std::bind(&PoseActionServerNode::handleCancel, this, _1),
+            std::bind(&PoseActionServerNode::handleAccepted, this, _1));
+}
+
+rclcpp_action::GoalResponse PoseActionServerNode::handleGoal(
+    const rclcpp_action::GoalUUID& uuid,
+    std::shared_ptr<const vortex_msgs::action::FilteredPose::Goal> goal) {
+    RCLCPP_INFO(this->get_logger(), "Received request to filter pose.");
+    (void)uuid;
+    if (is_executing_action_) {
+        RCLCPP_WARN(this->get_logger(),
+                    "Already executing an action, rejecting new goal.");
+        return rclcpp_action::GoalResponse::REJECT;
+    }
+    is_executing_action_ = true;
+    num_measurements_ = goal->num_measurements;
+    pose_queue_.clear();
+    return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
+}
+
+rclcpp_action::CancelResponse PoseActionServerNode::handleCancel(
+    const std::shared_ptr<GoalHandleFilteredPose> goal_handle) {
+    RCLCPP_INFO(this->get_logger(), "Received request to cancel goal");
+    (void)goal_handle;
+    is_executing_action_ = false;
+    active_goal_handle_.reset();
+    pose_queue_.clear();
+    return rclcpp_action::CancelResponse::ACCEPT;
+}
+
+void PoseActionServerNode::handleAccepted(
+    const std::shared_ptr<GoalHandleFilteredPose> goal_handle) {
+    RCLCPP_INFO(this->get_logger(), "Goal accepted. Processing poses.");
+    is_executing_action_ = true;
+    active_goal_handle_ = goal_handle;
+}
+
+void PoseActionServerNode::pose_callback(
+    const geometry_msgs::msg::PoseStamped::ConstSharedPtr pose_msg) {
+    if (!is_executing_action_)
+        return;
+
+    pose_queue_.push_back(*pose_msg);
+
+    geometry_msgs::msg::PoseStamped feedback_msg = *pose_msg;
+    auto feedback =
+        std::make_shared<vortex_msgs::action::FilteredPose::Feedback>();
+    feedback->current_pose = feedback_msg;
+    active_goal_handle_->publish_feedback(feedback);
+
+    if (pose_queue_.size() == num_measurements_) {
+        geometry_msgs::msg::PoseStamped filtered_pose;
+        filtered_pose.header = pose_msg->header;
+
+        std::vector<double> x_positions;
+        std::vector<double> y_positions;
+        std::vector<double> z_positions;
+        std::vector<double> yaw_angles;
+
+        for (const auto& pose : pose_queue_) {
+            x_positions.push_back(pose.pose.position.x);
+            y_positions.push_back(pose.pose.position.y);
+            z_positions.push_back(pose.pose.position.z);
+
+            double qz = pose.pose.orientation.z;
+            double qw = pose.pose.orientation.w;
+            double yaw = 2.0 * std::atan2(qz, qw);
+            yaw_angles.push_back(yaw);
+        }
+
+        std::sort(x_positions.begin(), x_positions.end());
+        std::sort(y_positions.begin(), y_positions.end());
+        std::sort(z_positions.begin(), z_positions.end());
```

Comment:
For the position part, you can use some neat C++20 features to combine all this:
```C++
    auto extract_and_sort_position = [&pose_queue_](auto position_extractor) {
        std::vector<double> sorted(
            (pose_queue_ | std::views::transform(position_extractor)).begin(),
            (pose_queue_ | std::views::transform(position_extractor)).end());
        std::ranges::sort(sorted);
        return sorted;
    };

auto x_positions = extract_and_sort_position([](const auto& pose) {
        return pose.pose.position.x;
    });
...
```
And similarly for yaw